### PR TITLE
fix(transactions): don't set transaction on continuation storage when…

### LINKED ISSUE #8429: Manually managed transactions set namespace value even when not executing under Sequelize._clsRun
## What are you doing?
My solution is a multi-tenant sequelize-based server, and nested transactions with CLS enabled are leaking between transactions.

Consider the following:

```js
const namespace = cls.createNamespace('tenant');
const metadataDatabase = new Sequelize(...);
const tenantDatabase = new Sequelize(...);

Sequelize.useCLS(namespace);

await tenantDatabase.transaction(async () => {
  const metadataTransaction = await metadataDatabase.transaction();
  
   await metadatabase.create({...}, { transaction: metadataTransaction });
   await metadataTransaction.commit();

   await tenantDatabase.create({...});
});
```

## What do you expect to happen?
In this example, the call to `tenantDatabase` will attempt to use the metadata's transaction as the outer one was overridden on the namespace by the manually managed transaction.  Being that `Sequelize._clsRun` is almost never executed to wrap a manually managed transaction, the issue hasn't presented itself.

In Sequelize, if the namespace exists the transaction is unconditionally put there, and immediately lost afterwards because `Sequelize._clsRun` is not used to bind the continuation context.  In this case, the outer managed transaction is running the continuation chain, hence we get queries that executte under another connection.

I am preparing a patch which does not attempt to override the continuation storage's transaction object when constructing a manual transaction, and associated tests now.

## What is actually happening?

The query for `tenantDatabase` is executed under the query for `metadataDatabase` erroneously.

__Dialect:__  any
__Dialect version:__ ???
__Database version:__ sqlite3
__Sequelize version:__ 4.13.5
__Tested with master branch:__ Yes

### PR DESCRIPTION
**Pull request by tylerjwatson**
… manually managed

### Pull Request check-list

_Please make sure to review and check all of these items:_

- [X] Does `npm run test` or `npm run test-DIALECT` pass with this change (including linting)?
- [X] Does the description below contain a link to an existing issue (Closes #[issue]) or a description of the issue you are solving?
- [X] Have you added new tests to prevent regressions?
- [ ] Is a documentation update included (if this change modifies existing APIs, or introduces new ones)?
- [X] Did you follow the commit message conventions explained in CONTRIBUTING.md?

### Description of change

This patch prevents Sequelize from adding a transaction to the continuation storage namespace when the transaction is manually managed.  Traditionally `Transaction#prepareEnvironment` would unconditionally set the transaction on the continuation storage, but is sunk immediately because manual transactions are never run under `Sequelize._clsRun`.  

Has associated test.  Closes #8429.

### CODE REVIEW

File: test/integration/cls.test.js [hunk 1]
```diff
@@ -33,6 +33,18 @@ if (current.dialect.supports.transactions) {
     });
 
     describe('context', () => {
+      it('does not use continuation storage on manually managed transactions', function() {
+        const self = this;
+
+        return Sequelize._clsRun(() => {
```
Comment (sushantdhiman) [inline test/integration/cls.test.js]: Fix indentation, its breaking build

File: test/integration/cls.test.js [hunk 2]
```diff
@@ -33,6 +33,18 @@ if (current.dialect.supports.transactions) {
...
+          return this.sequelize.transaction()
+          .then(transaction => {
+            console.error(self.ns.get('transaction'));
```
Comment (sushantdhiman) [inline test/integration/cls.test.js]: Remove this line

File: test/integration/cls.test.js [hunk 3]
```diff
@@ -33,6 +33,18 @@ if (current.dialect.supports.transactions) {
     });
 
     describe('context', () => {
+      it('does not use continuation storage on manually managed transactions', function() {
+        const self = this;
+
+        return Sequelize._clsRun(() => {
+          return this.sequelize.transaction()
+          .then(transaction => {
```
Comment (sushantdhiman) [inline test/integration/cls.test.js]: transaction is not used

**Comment by tylerjwatson** [comment]
Ah true, I'll fix that up shortly.

File: test/integration/cls.test.js (hunk 1 above)
Comment (tylerjwatson) [reply test/integration/cls.test.js]: I'm confused a bit here, the indentation looks fine.  Can you please point me to where the rules are being violated?

Comment (sushantdhiman) [reply test/integration/cls.test.js]: just run `npm run lint` you will know

File: lib/transaction.js [hunk 4]
```diff
@@ -99,7 +99,7 @@ class Transaction {
       });
   }
 
-  prepareEnvironment() {
+  prepareEnvironment(useCLS = true) {
```
Comment (sushantdhiman) [inline lib/transaction.js]: Oh, default variable is not supported in Node 4/6

**Comment by tylerjwatson** [comment]
I tested using the sqlite dialect fully and it appears its okay.  Perhaps some mssql specific stuff?

File: test/integration/cls.test.js [hunk 5]
```diff
@@ -33,6 +33,16 @@ if (current.dialect.supports.transactions) {
     });
 
     describe('context', () => {
+      it('does not use continuation storage on manually managed transactions', function() {
+        const self = this;
+
+        return Sequelize._clsRun(() => {
+          return this.sequelize.transaction().then(() => {
```
Comment (janmeier) [inline test/integration/cls.test.js]: Remember to roll back the transaction, I think that's what's causing some of the other tests to hang

Comment (tylerjwatson) [reply test/integration/cls.test.js]: Oooh, good point.  I'll fix that up right away.

**Comment by tylerjwatson** [comment]
I've never had to use `cls.get('transaction')`, as that would not be applicable for my use case anyway, I can't speak for everybody else.  If the `uuid` property exists publicly on a Sequelize instance, or, we make a `sequelize.clsTransaction` getter which will return the transaction from the storage with the proper key, I think nobody is going to be left out.

**Comment by janmeier** [comment]
You are probably right, hopefully CLS means you should never have to worry about the transaction

File: lib/transaction.js [hunk 6]
```diff
@@ -125,7 +129,7 @@ class Transaction {
         throw setupErr;
       }))
       .tap(() => {
-        if (this.sequelize.constructor._cls) {
+        if (useCLS && this.sequelize.constructor._cls) {
```
Comment (sushantdhiman) [inline lib/transaction.js]: `this.sequelize.constructor._cls` should not set if you dont call `useCls`, can you verify if your test fails without this fix ?

**Comment by tylerjwatson** [comment]
The patch works for my usecase and does indeed fix the immediate problem, but I do think it's worth investigating how to isolate transaction storage to sequelize instances.  The patch is still valid, but I think I can do better.  It will be some work.

**Comment by tylerjwatson** [comment]
Just as a side question, would you guys ever consider just pulling in `continuation-local-stoage` internally for the transaction context passing so it's always transparent and out of the way, and doing away with such APIs like `Sequelize.useCLS`?  We can allow people to override the namespace which is set if they want to, but it just uses a `sequelize` namespace by default?

**Comment by janmeier** [comment]
> Just as a side question, would you guys ever consider just pulling in continuation-local-stoage internally

Probably not:
* It's using an unofficial API (async-listener, https://github.com/node-forward/discussions/issues/28) - not that there are any better alternatives currently, but there might soon be with async-hooks https://github.com/Jeff-Lewis/cls-hooked
* It has a performance impact (async-listener, + having to patch bluebird to preserve context). So we only want to enable it if the user explicitly requests it.

**Comment by tylerjwatson** [comment]
Thought as much.  I'll continue as normal.

File: lib/transaction.js (hunk 6 above)
Comment (sushantdhiman) [reply lib/transaction.js]: Ok understood, I thought we were saving env without cls

**Comment by jeremija** [comment]
> It's using an unofficial API (async-listener, node-forward/discussions#28) - not that there are any better alternatives currently, but there might soon be with async-hooks https://github.com/Jeff-Lewis/cls-hooked

Hi @janmeier, could you share what's the reason against using [cls-hooked](https://github.com/Jeff-Lewis/cls-hooked) at the moment? Is it still too early and experimental? On the first try it seems to be working better than the original CLS since `async/await` expressions just work. This is related to #6834. Thanks!

**Comment by janmeier** [comment]
@jeremija I haven't experimented with it, you are welcome to do so and provide a PR :)

**Pull request merged**